### DISCUSSION TITLE
Auxkernel and postprocessor dependancy / run order

### DISCUSSION THREAD

**Discussion by errikosMFB** [idaholab/moose#25845] [Q&A General] [answered]
Hi all,

Below is the input file for a generic example where I'm incrementing an auxvariable field by a value found in a postprocessor.

I want to perform the following calculations on timestep_end, in order:
1.  Set the value of auxvariable "vaux" using an auxkernel
2. Calculate the average of vaux using a postprocessor ("vaux_avg")
3. Add the two in another auxkernel and save to "vaux2"

```
[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 1
    ymin = 0
    ymax = 1
    nx = 40
    ny = 40
    elem_type = QUAD4
  []
[]


[Variables]
  [u]
    # Dummy nonlinear system
    family = LAGRANGE
    order = FIRST
  []
[]

[Kernels]
  [null]
    type = NullKernel
    variable = u
  []
[]

[AuxVariables]
  [vaux]
    family = MONOMIAL
    order = CONSTANT
  []
  [vaux2]
    family = MONOMIAL
    order = CONSTANT
  []
[]

[AuxKernels]
  [k1]
    # Sets the value of variable vaux
    type = ConstantAux
    variable = vaux
    value = 0.1
    execute_on = 'timestep_end'
  []
  [k2coupledtopp]
    # Custom auxkernel. Simply couples to an auxvariable "vaux" and a postprocessor "vaux_pp" and returns the sum of the two
    type = TestPPAux
    variable = vaux2
    vaux = vaux
    vaux_pp = vaux_avg
    execute_on = 'timestep_end'
  []
[] # AuxKernels

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON' 
  # line_search = 'none'
  num_steps = 2
  dt = 1.0

  petsc_options_iname = '-pc_type -ksp_type'
  petsc_options_value = 'lu preonly'
[]

[Postprocessors]
  [vaux_avg]
    type = ElementAverageValue
    variable = vaux
    execute_on = 'timestep_end'
  []
  [vaux2_avg]
    type = ElementAverageValue
    variable = vaux2
    execute_on = 'timestep_end'
  []
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]


[Debug]
  show_execution_order = 'timestep_end'
[]
```

We expect vaux to have a value of 0.1 and vaux2 to have a value of 0.2. However, at the end of the first timestep, vaux2 holds a value of 0.1 (and vaux holds a value of 0.0!):

```
Postprocessor Values:
+----------------+----------------+----------------+
| time           | vaux2_avg      | vaux_avg       |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   1.000000e-01 |   0.000000e+00 |
+----------------+----------------+----------------+
```

The debug output for the timestep_end order is as follows:

```
[DBG] Computing elemental user objects on TIMESTEP_END
[DBG] Execution order of objects types on each element then its sides:
[DBG] - element user objects
[DBG] - domain user objects
[DBG] - element user objects contributing to the Jacobian
[DBG] - side user objects
[DBG] - domain user objects executing on sides
[DBG] - side user objects contributing to the Jacobian
[DBG] - internal side user objects
[DBG] - domain user objects executing on internal sides
[DBG] - interface user objects
[DBG] - domain user objects executing at interfaces
[DBG] Ordering of User Objects on block 0
[DBG] Executing ElementUserObject on TIMESTEP_END
[DBG] Order of execution:
[DBG] vaux_avg
                             
[DBG] Only user objects active on local element/sides are executed
[DBG] Executing auxiliary kernels on elements on TIMESTEP_END
[DBG] Ordering of AuxKernels on block 0
[DBG] k1 k2coupledtopp
[DBG] Computing elemental user objects on TIMESTEP_END
[DBG] Ordering of User Objects on block 0
[DBG] Executing ElementUserObject on TIMESTEP_END
[DBG] Order of execution:
[DBG] vaux2_avg
                             
[DBG] Only user objects active on local element/sides are executed
```
It seems that the postprocessor "vaux_avg" runs before the value of the variable "vaux" is set in the "k1" kernel. So it stores the initial value (0.0), and not the updated value (0.1).

I believe this is because in each "execution block" in moose, auxkernels run before postprocessors (as seen in the debug output above). Since an auxkernel depends on a PP value, a seperate block just with the PP is created and put before that one. Moose does not move the "k1" kernel to the first block as it does not see any dependancy between the postprocessor "vaux_avg" and the auxkernel "k1".

Is there any way for me to manipulate the behaviour to get the order that I want?

(I cannot use ICs to set the value of vaux, and this must all be done at timestep_end, because this is a simplification of the real code I'm working with, which has more complicated calculations than these).

Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

> Is there any way for me to manipulate the behaviour to get the order that I want?

yup that s what you're going to have to do.
I would recommend running:
vaux and vaux_avg on linear (or nonlinear)
vaux2 on timestep2

that way the first two are going to be updated all the time and vaux2 can come in at any point and get the right value. 

If you have multiapps there's other execute_on flags we can try.

Guillaume

**Reply by errikosMFB** [reply]
Hi Guillaume,

Thanks for the answer. I would like to not have to run vaux/vaux_avg on linear as they can be computationally expensive.

I do have multiapps. The main app does this and then passes vaux2 to a FullSolveMultiapp that executes on timestep_end. The subapp then updates some iterative field updates and transfers back to the main app.

Are there any flags between timestep_end and whenever the data gets transfered to the subapp that I could use? I am only aware of the ones found here : [https://mooseframework.inl.gov/source/interfaces/SetupInterface.html][https://mooseframework.inl.gov/source/interfaces/SetupInterface.html](https://mooseframework.inl.gov/source/interfaces/SetupInterface.html)

**Reply by GiudGiud** [reply]
ah no there's not. 
There's MULTIAPP_FIXED_POINT_END after TIMESTEP_END. 

You could do:
vaux on NONLINEAR
vaux_avg on TIMESTEP_END with force_preaux = true
vaux2 on TIMESTEP_END
